### DISCUSSION TITLE
wire:navigate shows "NULL" when navigating back x times.

### DISCUSSION THREAD

**Discussion by StoneyEagle** [livewire/livewire#6776] [7. Bugs (archived)]
### Livewire version

3.0.2

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

1. Add 2 pages, one empty, one dashboard view with a basic tailwind page skeleton moved to the layout.
1. Use wire:navigate(.hover) in your links.
1. Add x amount of DOM nodes and, the more you add the quicker the result.
1. Navigate a bunch of times between the pages, and go back in history with the back button or with `history.back()`.

### Additional information

When navigating back in history there is a point where Livewire renders `NULL` in the page for every page further back.

The amount of times you can go back is relative to the amount of DOM nodes spawned in the document by your components and layout.

I can navigate back 29 times before seeing the NULL.
When I add 30000 `<div class="bg-auto-100 "></div>` nodes, I can only go back once before it shows the NULL the next time I go back. (extreme example to test my theory)

Adding the following does not indicate anything that may be wrong. 
```js 
document.addEventListener('livewire:initialized', function() {
    window.addEventListener('popstate', function(event) {
        console.log(Alpine.history);
        console.log('history state change', event);
    });
});
```

When I do NOT use wire:navigate the navigate back works as expected.

P.S. My friend who also uses the latest version of Livewire has the same issue and was the one pointing this out to me.
Please let me know if there is anything unclear or need me to help.

Example: 
![image](https://github.com/livewire/livewire/assets/45034970/ef9468af-4e97-4899-8a00-57094cc298b9)
![image](https://github.com/livewire/livewire/assets/45034970/066400f2-f34d-4b4b-9201-48f7488f38e6)
![image](https://github.com/livewire/livewire/assets/45034970/6451016e-ea6e-432e-a511-f1899b14be5b)

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by Graphicfort** [comment]
I took a quick look at the code, and IDK if this is the correct answer.

### How navigate works:
When you visit a page using navigate the script will store the page data inside the browser Session Storage, and at some point the Session Storage will be full, and storing more data will throw an error 22 and will not store more data.

### How to store more data after error 22:
By deleting an old data and store a new one.

### Why null:
Because the requested data is gone.

### The solution:
IDK but may be by reloading the page before getting null message.

### Example:
- Visiting page 1, 2, 3,  and 4 (all 4 pages are stored in storage and the storage is full).
- Visiting page 5, will delete page 1 and may be 2 (depend on the number of characters on page 5).
- using history back button will return until page 2 and because page 1 was deleted, the null message will appear.

**Comment by StoneyEagle** [comment]
I have been looking through the code and came to the same conclusion.
I have made some code changes that solve this problem 50% but it is still unacceptable.
The way the history is managed will always result in the history being evicted, like for example when the user closes the browser and has tab restore enabled.
I have made the following changes to the code and it resulted in being able to store 900 pages in history before the browser ran out of memory.

```js
 function whenTheBackOrForwardButtonIsClicked(callback) {
    window.addEventListener("popstate", (e) => {
      let state = e.state || {};
      let alpine = state.alpine || {};
      if (!alpine._html)
        return;
      let html = fromSessionStorage(alpine._html);
      if (!html) { // catch returns null so we reload the page.
          window.location.reload();
          return;
      }
      callback(html);
    });
  }

  var historyStorage = []; // put the history in memory.
  function fromSessionStorage(timestamp) {
      try {
        return JSON.parse(historyStorage[timestamp]);
      } catch {
        return null;
      }
  }
  function tryToStoreInSession(timestamp, value) {
      historyStorage[timestamp] = JSON.stringify(value);
  }
  ```
  
I know that putting all the data in memory is not a good option but it is better than having only a few pages the user can go back to.
When the history is unavailable it will reload the page instead so there is at least no broken page for the end user.

I do believe this is at least a good temporary solution.

**Reply by Graphicfort** [reply]
using memory would make the browser crashes specially when using a mobile.

**Reply by StoneyEagle** [reply]
Yeah, limiting the number of pages stored to 50 or based on the total size could help.

**Comment by StoneyEagle** [comment]
What if we do something like this?

```js

  function whenTheBackOrForwardButtonIsClicked(callback) {
    window.addEventListener("popstate", async (e) => { // make async to enable fetch
      let state = e.state || {};
      let alpine = state.alpine || {};
      if (!alpine._html)
        return;
      let html = fromSessionStorage(alpine._html);
      if (!html) {
          try { // fetch the page from the backend.
              html = await fetch(location.href).then((i) => i.text());
          } catch { // and if it fails reload the page
              window.location.reload();
              return;
          }
      }
      callback(html);
    });
  }
```

We can keep the session storage.
This is a perfectly acceptable solution for me, however, disabling locally saved history prevents stale data.
So maybe it is a good idea to have a config variable that allows for disabling local history.

**Reply by Fill84** [reply]
I tested this, and this is working for me and i think it's the best solution so far for this issue. Now I don't get null pages anymore

**Reply by Graphicfort** [reply]
Perfect

**Reply by ArtMin96** [reply]
Hi @StoneyEagle. Where should this function be called?

**Comment by robsontenorio** [comment]
This issue happens when I have `laravel-debugbar` ON. So it injects a huge amount of html. **Turned off and issue solved.**

Also I can see `wire:navigate` cycles old pages, by looking at browser session storage when navigating  forwarding and backwarding.

<img width="720" alt="image" src="https://github.com/livewire/livewire/assets/118955/29827157-44ea-4460-a1fd-03da9046295e">

**Comment by Fill84** [comment]
```
function whenTheBackOrForwardButtonIsClicked(callback) {
    window.addEventListener("popstate", (e) => {
      let state = e.state || {};
      let alpine = state.alpine || {};
      if (!alpine._html)
        return;
      let html = fromSessionStorage(alpine._html);
      if (!html) {
        return fetchHtml(window.location, callback);
      }
      callback(html);
    });
  }
```

this uses the internal fetchHTML function with callback. I hope that you will implement this fix, because i'm getting sick of putting it in my self all the time after an update! please fix this!

**Reply by tominal** [reply]
I made this change in a fork, but I still get null on an iPhone after letting the page sit in a tab for a day or so. Maybe I’m misunderstanding how this is working. Should I just remove all wire:navigate.hover’s from my application? Or all wire:navigate’s?

**Reply by Fill84** [reply]
Then you didn't implement it correctly because it should fetch when html is null and return the original callback

**Reply by tominal** [reply]
As far as I know I implemented it correctly, rebuilt, pulled, then rebuilt on the frontend of the application. Then after a day or so of what I am assuming session-expiration in an iPhone tab, I see the page, then it immediately kills itself to null.

Here's what I updated in `js/plugins/navigate/history.js`:

```js
import { fetchHtml } from '@/plugins/navigate/fetch'

...

export function whenTheBackOrForwardButtonIsClicked(callback) {
    window.addEventListener('popstate', e => {
        let state = e.state || {}

        let alpine = state.alpine || {}

        if (! alpine._html) return

        let html = fromSessionStorage(alpine._html)

        if (!html)
            return fetchHtml(window.location, callback);

        callback(html)
    })
}

...

```

Thoughts?

**Reply by tominal** [reply]
I found removing all mention of wire:navigate attributes prevents that behavior thankfully.

**Comment by luanfreitasdev** [comment]
I tried implementing wire:navigate on some systems, and in some cases this still happens to me too.

**Comment by JakeOcean** [comment]
I noticed on my iPhone, if I force quite firefox, then reopen it, my Livewire compontent and the entire browser page just says "null" -- it happens 100% of the time I force quite firefox on iOS, but I havent been able to trigger it another way.  I am using wire.navigate.hover too

**Reply by tominal** [reply]
Ok so wire:navigate is not viable then. You can not release a production version of a project if you can guarantee the customer will see a blank page during revisit. My fork build also does nothing to remedy this issue: https://github.com/livewire/livewire/discussions/6776#discussioncomment-7438056 

So the question is - what are these dang iPhones doing that is completely breaking `wire:navigate.hover`?

Apparently there is a way to get to view the iPhone's inspect element console through a Macbook, but I have yet to try it. I am hoping there is a glaring error in the console so that we can push a fix to this thing.

**Comment by luanfreitasdev** [comment]
This still doesn't work on large applications. Clicking on some links (7 or so) and trying to go back, I either get null or Alpine crashes because it couldn't mount the livewire component. 

I still don't know how to reproduce the bug to send a "_Failing Test_" PR because the problem only happens when there is a lot of content in the HTML, and perhaps the memory is not being lost as everyone has reported. @calebporzio (_Sorry about that. Sometimes the solution may or may not be simple._)

This comment seems to solve it, but there is still a problem when restoring the Alpine's state
https://github.com/livewire/livewire/discussions/6776#discussioncomment-7438056

**Comment by lele85** [comment]
It's pretty easy to reproduce even just using the livewire docs:

https://livewire.laravel.com/docs

Just navigate around 15 times the various sections:

![Screenshot 2024-01-30 alle 13 45 11](https://github.com/livewire/livewire/assets/49144/2b9dcaee-99a7-4ee8-8eda-ac740ec5788e)

Then start to hit the back button for 11 to 13 times and 💥 :

![Screenshot 2024-01-30 alle 13 46 43](https://github.com/livewire/livewire/assets/49144/c2265169-1d7b-405f-9f60-96535f5ae54a)

For now i patched adding the fallback proposed by @Fill84 and it solves the problem but in my opinion there are still issues:

- the loading indicator is not working when we have a "cache miss"
- the strategy of filling the `sessionStorage` until it breaks is a bit risky because it will probably happen, and third party libraries may often find the session storage full

Ideally the caching behavior should be:
- Easely switched off
- Have a sensitive fallback strategy
- Rely on a sliding window that is not the dimension of the whole session storage.

The other problem that the current back mechanism has is that if you rely on:

```
@script
@endscript
``` 

to fire some JS when a page is navigated, when you it the cache this will never happen.

**Reply by luanfreitasdev** [reply]
Thanks @lele85. Well noted

**Reply by luanfreitasdev** [reply]
Is there a chance to use IndexedDB?

**Reply by lele85** [reply]
@luanfreitasdev I think the problem is not the underlying db uses, plus `IndexedDB` is async and has quite a terrible API, in general the caching strategy that is implemented now just doesn't play well in a lot of edge cases. Honestly a good first step would be to just have a switch to disable completely the behavior. That said i understand if for now this is not a priority. Maybe @calebporzio can share some thoughts on this?

**Comment by joshhanley** [comment]
Caleb is currently working on this in PR #7895

**Comment by StoneyEagle** [comment]
I have tested it and the history looks to be fixed, much appreciated!
